Improve handler's `process_request` in cluster

mhamra opened the pull request:
|Related issue|
|---|
|[16509](https://github.com/wazuh/wazuh/issues/16509)|

## Description

Optimize handler.process_request function. Refactor process_request function to lookup in a dictionary instead of using if/elif/else statements to process request commands.

Selutario commented on framework/wazuh/core/cluster/local_client.py:
```diff
@@ -18,6 +18,22 @@ class LocalClientHandler(client.AbstractClient):
     """
 
+    def _create_cmd_handlers(self):
+        """
+            Add handlers to _cmd_handler dictionary
+        """
```
```suggestion
        """Add handlers to _cmd_handler dictionary.
        
        """

```

mhamra replied:
Fixed Docstring

Selutario commented on framework/wazuh/core/cluster/local_client.py:
```diff
@@ -46,6 +62,41 @@ def connection_made(self, transport):
     def _cancel_all_tasks(self):
         pass
 
+    def _cmd_dapi_res_or_send_f_res(self, data: bytes) -> Tuple[bytes, bytes]:
+        if data.startswith(b'Error'):
+            return b'err', self.process_error_from_peer(data)
+        elif data not in self.in_str:
+            return b'err', self.process_error_from_peer(b'Error receiving string: ID ' + data + b' not found.')
+        self.response = self.in_str[data].payload
+        self.response_available.set()
+        # Remove the string after using it
+        self.in_str.pop(data, None)
+        return b'ok', b'Distributed api response received'
+
+    def _cmd_ok(self, data: bytes) -> Tuple[bytes, bytes]:
+        if data.startswith(b'Error'):
+            return b'err', self.process_error_from_peer(data)
+        self.response = data
+        self.response_available.set()
+        return b'ok', b'Sendsync response received'
+
+    def _cmd_control_res(self, data: bytes) -> Tuple[bytes, bytes]:
+        if data.startswith(b'Error'):
+            return b'err', self.process_error_from_peer(data)
+        self.response = data
+        self.response_available.set()
+        return b'ok', b'Response received'
+
+    def _cmd_dapi_err(self, data: bytes) -> Tuple[bytes, bytes]:
+        self.response = data
+        self.response_available.set()
+        return b'ok', b'Response received'
+
+    def _cmd_err(self, data: bytes) -> Tuple[bytes, bytes]:
+        self.response = data
+        self.response_available.set()
+        return b'ok', b'Error response received'
```
Missing docstrings

Selutario commented on framework/wazuh/core/cluster/local_server.py:
```diff
@@ -363,7 +372,44 @@ class LocalServerHandlerWorker(LocalServerHandler):
+        )
+
+    # Start command Handlers
+    def _cmd_sendsync(self, data: bytes):
+        if self.server.node.client is None:
+            raise WazuhClusterError(3023)
+        asyncio.create_task(self.log_exceptions(
+            self.server.node.client.send_request(b'sendsync', self.name.encode() + b' ' + data)))
+        return None, None
+
+    def _cmd_sendasync(self, data: bytes):
+        if self.server.node.client is None:
+            raise WazuhClusterError(3023)
+        asyncio.create_task(self.log_exceptions(
+            self.server.node.client.send_request(b'sendsync', self.name.encode() + b' ' + data)))
+        return b'ok', b'Added request to sendsync requests queue'
+
+    def _cmd_dapi(self, data: bytes):
+        if self.server.node.client is None:
+            raise WazuhClusterError(3023)
+        asyncio.create_task(self.log_exceptions(
+            self.server.node.client.send_request(b'dapi', self.name.encode() + b' ' + data)))
+        return b'ok', b'Added request to API requests queue'
```
Missing docstrings

mhamra replied:
Added Docstrings to all new methods

Selutario commented on framework/wazuh/core/cluster/master.py:
```diff
@@ -227,6 +256,42 @@ def to_dict(self):
                            'last_keep_alive': self.last_keepalive}
                 }
 
+    def _cmd_syn_w_g_e(self, data: bytes):
+        logger = self.task_loggers['Agent-groups send']
+        start_time = self.send_agent_groups_status['date_start']
+        return c_common.end_sending_agent_information(logger, start_time, data.decode())
+
+    def _cmd_syn_wgc_e(self, data: bytes):
+        logger = self.task_loggers['Agent-groups send full']
+        start_time = self.send_agent_groups_status['date_start']
+        return c_common.end_sending_agent_information(logger, start_time, data.decode())
+
+    def _cmd_syn_w_g_err(self, data: bytes):
+        logger = self.task_loggers['Agent-groups send']
+        return c_common.error_receiving_agent_information(logger, data.decode(), info_type='agent-groups')
+
+    def _cmd_syn_wgc_err(self, data: bytes):
+        logger = self.task_loggers['Agent-groups send full']
+        return c_common.error_receiving_agent_information(logger, data.decode(), info_type='agent-groups')
+
+    def _cmd_dapi(self, data: bytes):
+        self.server.dapi.add_request(self.name.encode() + b'*' + data)
+        return b'ok', b'Added request to API requests queue'
+
+    def _cmd_get_nodes(self, data: bytes):
+        cmd, res = self.get_nodes(json.loads(data))
+        return cmd, json.dumps(res).encode()
+
+    def _cmd_get_health(self, data: bytes):
+        cmd, res = self.get_health(json.loads(data))
+        return cmd, json.dumps(
+            res, default=lambda o: "n/a" if isinstance(o, datetime) and o == utils.get_date_from_timestamp(0) else
+            (o.__str__() if isinstance(o, datetime) else None)).encode()
+
+    def _cmd_sendsync(self, data: bytes):
+        self.server.sendsync.add_request(self.name.encode() + b'*' + data)
+        return b'ok', b'Added request to SendSync requests queue'
+
```
Missing docstring, review this in all newly added methods.

mhamra replied:
Added Docstrings to all new methods

Selutario commented on framework/wazuh/core/cluster/server.py:
```diff
@@ -17,15 +17,28 @@
     """
 
+    def _create_cmd_handlers(self):
+        """
+            Add command handlers to _cmd_handler dictionary
+        """
```
Review this everywhere https://numpydoc.readthedocs.io/en/latest/format.html#short-summary

Also, move it after `__init__` method in all the modules.

mhamra replied:
Moved after __init__ and fixed Docstring

Selutario reviewed (changes requested):
In the future, please avoid including changes not related to the initial scope of the issue (format, line breaks, etc). A relatively small PR ends up being much larger to review.

mhamra commented on framework/wazuh/core/cluster/client.py:
```diff
@@ -106,31 +109,46 @@ async def start(self):
+            {
+                b'echo-m': lambda _, data: self.echo_client(data),
+            }
+        )
```
Moved after __init__

mhamra commented on framework/wazuh/core/cluster/local_client.py:
```diff
@@ -18,6 +18,22 @@ class LocalClientHandler(client.AbstractClient):
+                b'dapi_err': lambda _, data: self._cmd_dapi_err(data),
+                b'err': lambda _, data: self._cmd_err(data),
+            }
+        )
```
Moved after __init__

mhamra commented on framework/wazuh/core/cluster/master.py:
```diff
@@ -168,6 +168,34 @@ class MasterHandler(server.AbstractServerHandler, c_common.WazuhCommon):
+                b'sendsync': lambda _, data: self._cmd_sendsync(data),
+            }
+        )
+
```
Moved after __init__

mhamra commented on framework/wazuh/core/cluster/client.py:
```diff
@@ -106,31 +109,46 @@ async def start(self):
+    def _create_cmd_handlers(self):
+        """
+            Add command handlers to _cmd_handler dictionary
+        """
```
Fixed Docstring

mhamra commented on framework/wazuh/core/cluster/local_server.py:
```diff
@@ -25,6 +26,23 @@ class LocalServerHandler(server.AbstractServerHandler):
+            Assign _cmd_handler class dictionary copying super class 
+            _cmd_handler dictionary entries and adding command 
+            handlers
+        """
```
Fixed Docstring

mhamra commented on framework/wazuh/core/cluster/local_server.py:
```diff
@@ -41,36 +59,9 @@ def connection_made(self, transport):
-            return super().process_request(command, data)
+    def _cmd_send_file(self, data: bytes):
+        path, node_name = data.decode().split(' ')
+        return self.send_file_request(path, node_name)
```
Added Docstrings to all new methods

mhamra commented on framework/wazuh/core/cluster/local_server.py:
```diff
@@ -238,7 +231,35 @@ class LocalServerHandlerMaster(LocalServerHandler):
+            Assign _cmd_handler class dictionary copying super class 
+            _cmd_handler dictionary entries and adding command 
+            handlers
+        """
```
Fixed Docstring

mhamra commented:
### UPDATE
Rebase to master done. 

Fixed all changes suggested by @Selutario

Selutario reviewed (changes requested):
I'm getting this error in the master's `ossec.log` file when workers try to connect (`hello` command):
```
2023/05/26 09:03:41 INFO: [Worker] [Main] Connection from ('172.18.0.5', 38858)
2023/05/26 09:03:41 ERROR: [Worker] [Main] Unhandled error processing request 'b'hello'': cannot unpack non-iterable NoneType object
Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.6.0-py3.9.egg/wazuh/core/cluster/common.py", line 841, in dispatch
    Received data from other peer.
TypeError: cannot unpack non-iterable NoneType object
2023/05/26 09:03:41 ERROR: [Worker] [Main] Error during handshake with incoming connection.
```

mhamra commented:
### UPDATE

- git Rebase master of the branch
- Fixed AbstractServerHandler._create_cmd_handlers based on @Selutario review
- Fixed related unit tests
- Improved related unit tests

mhamra commented:
### UPDATE
- Fixed missing endpoints in Docstring's short descriptions.
- Fixed command_not_found tests
- Fixed process_request docstrings
- Removed _create_cmd_handlers duplicated before __init__

mhamra commented:
### UPDATE
The branch was rebased, and commits were reorganized for clarity

fdalmaup commented:
The cluster is being removed in 5.0 #27434 so no improvements are required.
